extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js

```diff
@@ -171,226 +113,100 @@ describe('amp-ad-3p-impl', () => {
-  event fired at least the given number of times.
-- errors: Array of console.error fired during page load.
-*/
-// Please note undefined and null are different for opt_canonical
-function getAd(opt_attributes, opt_canonical, opt_handleElement,
-    opt_beforeLayoutCallback) {
-  const attributes = opt_attributes || {
-    width: 300,
-    height: 250,
-    type: '_ping_',
-    src: 'https://testsrc',
-    'data-valid': 'true',
-  };
-  let canonical = opt_canonical;
-  if (opt_canonical === undefined) {
-    canonical = 'https://schema.org';
-  }
-  return createIframePromise(undefined, opt_beforeLayoutCallback)
-      .then(iframe => {
-        iframe.iframe.style.height = '400px';
-        iframe.iframe.style.width = '400px';
-        markElementScheduledForTesting(iframe.win, 'amp-user-notification');
-        if (canonical) {
-          const link = createElementWithAttributes(iframe.doc, 'link', {
-            rel: 'canonical',
-            href: canonical,
-          });
-          iframe.doc.head.appendChild(link);
+  describe('preconnectCallback', () => {
+    it('should add prefeches', done => {
+      ad3p.buildCallback();
+      ad3p.preconnectCallback();
+      setTimeout(() => {
+        let fetches = win.document.querySelectorAll('link[rel=prefetch]');
+        if (!fetches.length) {
+          fetches = win.document.querySelectorAll('link[rel=preload]');
         }
-        let a = createElementWithAttributes(iframe.doc, 'amp-ad', attributes);
+        expect(fetches).to.have.length(2);
+        expect(fetches[0]).to.have.property('href',
+            'http://ads.localhost:9876/dist.3p/current/frame.max.html');
+        expect(fetches[1]).to.have.property('href',
+            'http://ads.localhost:9876/dist.3p/current/integration.js');
+
+        const preconnects =
+            win.document.querySelectorAll('link[rel=preconnect]');
+        expect(preconnects[preconnects.length - 1]).to.have.property('href',
+            'https://testsrc/');
+        done();
+      }, 0);
+    });
+  });
 
-        // Make document long.
-        a.style.marginBottom = '1000px';
-        if (opt_handleElement) {
-          a = opt_handleElement(a);
-        }
-        return iframe.addElement(a);
-      });
-}
+  // TODO: add test for noContentHandler_()
 
-function getAdInAdContainer() {
-  return createIframePromise().then(iframe => {
-    const adContainer = createElementWithAttributes(iframe.doc,
-        'amp-sticky-ad', {layout: 'nodisplay'});
-    const ampAd = createElementWithAttributes(iframe.doc, 'amp-ad', {
-      width: 300,
-      height: 50,
-      type: '_ping_',
-      src: 'testsrc',
-      'data-valid': 'true',
-    });
-    const link = createElementWithAttributes(iframe.doc, 'link', {
-      rel: 'canonical',
-      href: 'blah',
+  describe('renderOutsideViewport', () => {
+    it('should allow rendering within 3 viewports by default', () => {
+      expect(ad3p.renderOutsideViewport()).to.equal(3);
```

Comment:
oh, i think i get it, because we never layout ad in this case. But then should we test that renderOutsideViewport will be false for 1 sec after we layout an ad?